cirq/google/calibration/phased_fsim.py

```diff
@@ -130,3 +140,200 @@ def override_by(self, other: 'PhasedFSimCharacterization') -> 'PhasedFSimCharact
             other that are not None). Otherwise the current values are used.
         """
         return other.merge_with(self)
+
+
+@dataclasses.dataclass(frozen=True)
+class PhasedFSimCalibrationResult:
+    """The PhasedFSimGate characterization result.
+
+    Attributes:
+        parameters: Map from qubit pair to characterization result. For each pair of characterized
+            quibts a and b either only (a, b) or only (b, a) is present.
+        gate: Characterized gate for each qubit pair.
+    """
+
+    parameters: Dict[Tuple[Qid, Qid], PhasedFSimCharacterization]
+    gate: Gate
+
+    def get_parameters(self, a: Qid, b: Qid) -> Optional['PhasedFSimCharacterization']:
+        """Returns parameters for a qubit pair (a, b) or None when unknown."""
+        if (a, b) in self.parameters:
+            return self.parameters[(a, b)]
+        elif (b, a) in self.parameters:
+            return self.parameters[(b, a)].parameters_for_qubits_swapped()
+        else:
+            return None
+
+    @classmethod
+    def _create_parameters_dict(
+        cls,
+        parameters: List[Tuple[Qid, Qid, PhasedFSimCharacterization]],
+    ) -> Dict[Tuple[Qid, Qid], PhasedFSimCharacterization]:
+        """Utility function to create parameters from JSON.
+
+        Can be used from child classes to instantiate classes in a _from_json_dict_
+        method."""
+        return {(q_a, q_b): params for q_a, q_b, params in parameters}
+
+    @classmethod
+    def _from_json_dict_(
+        cls,
+        **kwargs,
+    ) -> 'PhasedFSimCalibrationResult':
+        """Magic method for the JSON serialization protocol.
+
+        Converts serialized dictionary into a dict suitable for
+        class instantiation."""
+        del kwargs['cirq_type']
+        kwargs['parameters'] = cls._create_parameters_dict(kwargs['parameters'])
+        return cls(**kwargs)
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        """Magic method for the JSON serialization protocol."""
+        return {
+            'cirq_type': 'PhasedFSimCalibrationResult',
+            'gate': self.gate,
+            'parameters': [(q_a, q_b, params) for (q_a, q_b), params in self.parameters.items()],
+        }
+
+
+# We have to relax a mypy constraint, see https://github.com/python/mypy/issues/5374
+@dataclasses.dataclass(frozen=True)  # type: ignore
+class PhasedFSimCalibrationRequest(abc.ABC):
+    """Description of the request to characterize PhasedFSimGate.
+
+    Attributes:
+        pairs: Set of qubit pairs to characterize. A single qubit can appear on at most one pair in
+            the set.
+        gate: Gate to characterize for each qubit pair from pairs. This must be a supported gate
+            which can be described cirq.PhasedFSim gate.
+    """
+
+    pairs: Tuple[Tuple[Qid, Qid], ...]
+    gate: Gate  # Any gate which can be described by cirq.PhasedFSim
```

Comment:
What if `gate` is `ISwapPowGate` and you set `characterize_zeta` to `True`? How does this `gate` actually change the behavior of the algorithm?

Reply:
`gate` is just a selector of which hardware gate should be used for characterization. Our processor have a limited choice of those gates.

Reply:
So this can basically be `cirq.ISWAP**0.5` or `cg.SYC`? Maybe `cirq.ISWAP` as well? And it is used in construction of the moment calibration circuits. Does it also seed the algorithm with the correct "ideal" values for the angles?

Reply:
Also `cirq.CZ`. It is used in the construction for the moments and seeds the algorithm with the initial guess, that's correct.